File: pset6/sentimental-credit/credit.py
# TODO
from cs50 import get_string
import re


def main():
    number = get_string("What's your credit number? ")
    if checksum(number) == False:
        print("INVALID")
        quit()
    else:
        print('checksum ok')

    if numbers(number) == 'visa':
        print('VISA')
    elif numbers(number) == "mc":
        print('MASTERCARD')
    elif numbers(number) == 'amex':
        print('AMEX')
    else:
        print('INVALID')


# Check length and first numbers
def numbers(number):
    if re.search("^3(?:4|7).{13}$", number):
        return "amex"
    elif re.search("^5[1-5].{14}$", number):
        return "mc"
    elif re.search("^4.{12}$", number) or re.search("^4.{15}$", number):
        return "visa"


# Checksum
def checksum(number):
    sum = 0
    leng = len(number)

    # If number of digits is even
    for i in (range(leng - 1, 0, -2)):
        # Odd digits from last
        sum += int(number[i])

        # Even digits from last
        sraka = int(number[i - 1]) * 2
        if (sraka) > 9:
            zhopa = (sraka) % 10 + (sraka) // 10
            sum += zhopa
        else:
            sum += sraka

    # If number of digits is uneven we add one digit
    if leng % 2 != 0:
        # get last number in the array
        sum += int(number[0])

    # Test if checksum is alright
    if sum % 10 == 0:
        return True
    else:
        return False


main()

File: pset1/credit/credit.c
 #include <cs50.h>
#include <stdio.h>

int main(void)
{

    long number = 0;

    // Prompt to type in a credit card number
    do
    {
        number = get_long("What's your credit card number (without spaces)? ");
    }
    while (number < 1);

    // Calculate the checksum and get number length
    long num = number;
    int length = 0;
    int checksum = 0;
    int remainder = 0;

    do
    {
        remainder = num % 10;
        // Add every other digit form 1st right to checksum
        if (length % 2 == 0)
        {
            checksum = checksum + remainder;
        }
        // Add summarized digits of (every other digit form 2nd right * 2) to checksum
        // If (every other digit form 2nd right * 2) < 10 then it adds number + 0, no need for other conditional
        else
        {
            checksum = checksum + remainder * 2 / 10 + remainder * 2 % 10;
        }
        num = num / 10;
        length++;
    }
    while (num > 0);

    // Decide whether it's a valid card or not
    // Get first 2 digits
    long digits = number;
    
    do
    {
        digits = digits / 10;
    }
    while (digits > 100);

    // Check checksum last digit
    if (checksum % 10 == 0)
    {
        // If amex
        if ((digits == 34 || digits == 37) && length == 15)
        {
            printf("AMEX\n");
        }
        // Else if master
        else if ((digits > 50 && digits < 56) && length == 16)
        {
            printf("MASTERCARD\n");
        }
        // Else if visa
        else if ((length == 13 || length == 16) && digits / 10 == 4)
        {
            printf("VISA\n");
        }
        //else when checksum last digit is 0
        else
        {
            printf("INVALID\n");
        }
    }
    // If checksum lasst digit isn't 0
    else
    {
        printf("INVALID\n");
    }
}